### PR TITLE
feat: password with visibility toggle

### PR DESCRIPTION
**Pull request by Benehiko**
## Related Issue or Design Document

Although this looks simple to achieve on the surface, this has many hidden problems when implemented with CSS only.
The Account Experience only uses CSS with server-side rendering and has no support for JS on the client. This requires some more investigation into the viability of the feature with CSS-only and if it makes sense to persue this for the Account Experience at the current moment. 

When using CSS only we need to set the input field to [type=text] and set the password styling using `-webkit-security-text`. https://caniuse.com/mdn-css_properties_-webkit-text-security
This css property is not a standard and has only recently received support in firefox. There is a risk that users running on an obscure browser or using an old browser will always see their password with the toggle having no effect.

Some alternatives to consider:

1. Add a flag that allows users consuming the library to set this (css only method)
2. Add react support to swap out the input type using JS.

https://github.com/ory/elements/issues/116

image

## Checklist

- [ ] I have read the [contributing guidelines](../blob/master/CONTRIBUTING.md) and signed the CLA.
- [ ] I have referenced an issue containing the design document if my change introduces a new feature.
- [ ] I have read the [security policy](../security/policy).
- [ ] I confirm that this pull request does not address a security vulnerability. 
      If this pull request addresses a security vulnerability, 
      I confirm that I got green light (please contact [security@ory.sh](mailto:security@ory.sh)) from the maintainers to push the changes.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [ ] I have added necessary documentation within the code base (if appropriate).

### CODE REVIEW

**Comment by LBBO** [comment]
I really don't like having an insecure default for older browsers. How about adding two basically identical inputs, one with `type="password"` and one with `type="text"`, have the `text` hidden by default and then use a `@supports` query to show the `text` input and hide the `password` input iff this feature is supported? I would assume that browsers that support `@supports` are a superset of those that support `-webkit-security-text` and this way, if a browser doesn't support either of the features we still have secure defaults and the user just misses out on the "view password" feature.

**Comment by LBBO** [comment]
Hmm, while aparently most browsers added support for `@supports` way later than for `webkit-text-security`, it's the other way around for Firefox. And the percentages of supporting browsers are similar enough that I still think this would be a good idea.

**Comment by Benehiko** [comment]
Thank you for the input! I agree with this sentiment.

**Comment by Benehiko** [comment]
NextJS is failing to build due to the `@supports(:has)` css selector. The reason for this is NextJs bundles postcss and autoprefixer by default. There was a bug in autoprefixer which prevents it from understanding the `@supports` css selector with the pseudo parameter  `:has`. See this issue https://github.com/postcss/autoprefixer/issues/1391

So according to the autoprefixer issue, this should be fixed by now in version 10.2.5, but on version 10.4.x this is still a problem. 

I have added autoprefixer to vite so we can catch this on build time of our react components in the latest commit. I think we should also add a browser support list to the `.browserlistrc` file so that future css selectors added to components do not break the browsers we want to support.

```shell
 [vite:css] [postcss] Unknown node type in node creation
       file: /home/benehiko/Github/elements/src/theme/input-field.css.ts.vanilla.css:undefined:undefined
       error during build:
       Error: Unknown node type in node creation
           at /home/benehiko/Github/elements/src/theme/input-field.css.ts.vanilla.css:115:1
           at Rule.normalize (/home/benehiko/Github/elements/node_modules/postcss/lib/container.js:212:13)
           at Rule.append (/home/benehiko/Github/elements/node_modules/postcss/lib/container.js:30:24)
           at Supports.virtual (/home/benehiko/Github/elements/node_modules/autoprefixer/lib/supports.js:297:10)
           at Supports.prefixed (/home/benehiko/Github/elements/node_modules/autoprefixer/lib/supports.js:179:21)
           at /home/benehiko/Github/elements/node_modules/autoprefixer/lib/supports.js:35:29
           at Array.map (<anonymous>)
           at Supports.add (/home/benehiko/Github/elements/node_modules/autoprefixer/lib/supports.js:33:18)
           at Supports.process (/home/benehiko/Github/elements/node_modules/autoprefixer/lib/supports.js:231:16)
           at /home/benehiko/Github/elements/node_modules/autoprefixer/lib/processor.js:67:27
           at /home/benehiko/Github/elements/node_modules/postcss/lib/container.js:324:18
       npm ERR! Lifecycle script `build` failed with error:
       npm ERR! Error: command failed
       npm ERR!   in workspace: @ory/elements@0.0.0
       npm ERR!   at location: /home/benehiko/Github/elements/packages/react
```

@LBBO maybe you could take a look at this when you get a chance?

File: src/theme/input-field.css.ts [hunk 1]
```diff
@@ -39,3 +47,188 @@ export const inputFieldStyle = style({
...
+ * If both the :has selector and the -webkit-text-security: disc are supported,
+ * we hide this input field in favor of the custom password input field
+ **/
+globalStyle(`${inputFieldStyle}[type="password"]`, {
+  "@supports": {
+    "(-webkit-text-security: disc) and selector(:has(a,b))": {
+      display: "none",
+    },
+  },
+})
```
Comment (LBBO) [inline src/theme/input-field.css.ts:234]: Not sure how to test this properly, but to my understanding, selectors that can't be understood are just not matched. Therefore, we don't need a `@supports` clause for `:has` and can instead just use the selector, so long as the fallback is well-setup and secure. Perhaps something like this?

```suggestion
/**
 * Firefox does not support the :has selector yet, so we need to test for this
 * https://caniuse.com/css-has
 * https://caniuse.com/mdn-css_properties_-webkit-text-security
 * If both the :has selector and the -webkit-text-security: disc are supported,
 * we can show the password input field with a visibility toggle
 * and hide the fallback password input field
 **/
globalStyle(`${passwordInputContainerStyle}:has(*)`, {
  "@supports": {
    "(-webkit-text-security: disc)": {
      display: "flex",
    },
  },
})

/**
 * Firefox does not support the :has selector yet, so we need to test for this
 * https://caniuse.com/css-has
 * https://caniuse.com/mdn-css_properties_-webkit-text-security
 * If both the :has selector and the -webkit-text-security: disc are supported,
 * we hide this input field in favor of the custom password input field
 **/
globalStyle(`${inputFieldStyle}[type="password"]:has(*)`, {
  "@supports": {
    "(-webkit-text-security: disc)": {
      display: "none",
    },
  },
})
```

**Comment by LBBO** [comment]
I pushed some changes, let me know what you think!

**Comment by LBBO** [comment]
Tests are failing due to duplicate IDs with the two inputs now. Ideas:

a) only render the fallback and then switch using JS if the required features are supported
b) turn the entire wrapper into a label so we don't need the htmlFor anymore. might cause SEO issues (and rightfully so) and isn't really any better than having the ID clash
c) disambiguate the IDs using prefixes / postfixes & add another label that is shown / hidden with its input. would break any hard ID-references that current users have

File: src/theme/input-field.css.ts [hunk 2]
```diff
@@ -1,41 +1,211 @@
...
+  display: "none",
+  userSelect: "none",
+  MozUserSelect: "none",
+  WebkitUserSelect: "none",
+})
+
+// this is a label attached to the checkbox that toggles the visibility of the input field
+// we don't want to show the checkbox, but we want to show the label
+export const inputFieldVisibilityToggleLabelStyle = style({
+  cursor: "pointer",
```
Comment (mszekiel) [inline src/theme/input-field.css.ts:140]:
```suggestion
  display: "inline-flex",
  cursor: "pointer",
```

**Review by mszekiel — APPROVED** [review]

**Pull request merged**